flask-version/metagenomongo/app.py

```diff
@@ -113,70 +133,70 @@ def empty_check(last_data):
+    data_validation.validation_all( fields, options, errors, data)
     return render_template('index_with_table.html', \
                 tables=[data.to_html(classes='data', header="true")], \
-                fields=fields, results=results, values=values, \
+                errors=errors, \
                 df=data, user_name=request.form["user_name"])
 
 @app.route('/addLine', methods=['POST'])
 def addLine():
     data_list = parse_form_data(request.form)
-    results = []
-    values = {"default": 0}
+    errors = defaultdict(list)
+    email.email_env_check(errors)
     data = pd.DataFrame(data_list, columns=fields)
-    data_validation.validation_all( fields, options, results, data)
+    data_validation.validation_all( fields, options, errors, data)
     new_data = data.iloc[-1]
     if empty_check(new_data):
         data.loc[len(data)] = new_data
         data.at[len(data)-1,'sampleID'] = ''
     return render_template('index_with_table.html', \
                 tables=[data.to_html(classes='data', header="true")], \
-                fields=fields, results=results, values=values, \
+                errors=errors, \
                 df=data, user_name=request.form["user_name"])
 
 @app.route('/save', methods=['GET', 'POST'])
 def save():
     data_list = parse_form_data(request.form)
     user_name = request.form["user_name"]
-    values = {"default": 0}
-    results = []
-    if not check_user(user_name):
-        results.append({'error':'unauthorized user. Please contact the database admin'})
-        data = pd.DataFrame(data_list, columns=fields)
-        return render_template('index_with_table.html', \
-                    tables=[data.to_html(classes='data', header="true")], fields=fields, results=results, values=values, df=data)
+    errors = defaultdict(list)
+    email.email_env_check(errors)
     df = pd.DataFrame(data_list, columns=fields)
     df = df.loc[:, ~df.columns.str.contains('^Unnamed')]
-    data_validation.validation_all( fields, options, results, df)
-    if results:
+    data_validation.validation_all( fields, options, errors, df)
+    if errors['fatal_error']:
         return render_template('index_with_table.html', \
             tables=[df.to_html(classes='data', header="true")], \
-            fields=fields, results=results, values=values, \
+            errors=errors, \
             df=df, user_name=request.form["user_name"])
     output = io.StringIO()
     df.to_csv(output, index=False)
     mem = io.BytesIO()
     mem.write(output.getvalue().encode('utf-8'))
     mem.seek(0)
-    email.send_email() 
     current_time = datetime.datetime.now()
     file_name = user_name + current_time.strftime('_%Y-%m-%d-%H-%M-%S') +".csv"
-    save_file_server(output.getvalue(),file_name)
+    save_file_server(output.getvalue(),file_name,errors)
     return send_file(mem, mimetype='text/csv', \
                      as_attachment=True, download_name=file_name)
 
 @app.route('/', methods=['GET', 'POST'])
 def index():
```

Comment:
@iwmstjp I think this is a valid point, I thought about this myself when I read the `index` function. Consider this as you keep working on the application. This isn't urgent, what's important is that the application should work correctly, but at some point it might indeed make sense to reorganise the code by splitting it up into somewhat smaller chunks that have a well-defined goal and interface.